packages/asset-value/src/assetValue.ts

```diff
@@ -0,0 +1,212 @@
+import { Md5 } from 'ts-md5'
+
+import { CHECKSUM_LENGTH, DEFAULT_FORMAT_DECIMALS } from './constants'
+import { AssetValueParams, isAssetValueParams, SerializedAssetValue } from './types'
+import { AssetValueFormat } from './types'
+import { bn, bnOrZero } from './utils'
+
+type AssetValueState = {
+  assetId: AssetId
+  precision: number
+  value: string
+}
+
+//TODO(pastaghost): Add JSDoc documentation
+export class AssetValue {
+  state: AssetValueState
+
+  public constructor(params: AssetValueParams | SerializedAssetValue) {
+    if (isAssetValueParams(params)) {
+      if (params.asset && !(params.asset.assetId?.length && params.asset.precision > -1)) {
+        throw new Error('Cannot initialize AssetValue with invalid asset')
+      }
+
+      if (!params.asset && !(params.assetId?.length && params.precision > -1)) {
+        throw new Error('Cannot initialize AssetValue with invalid asset')
+      }
+
+      let _value: string
+      switch (params.format) {
+        case AssetValueFormat.BASE_UNIT:
+          _value = bnOrZero(params.value).toFixed(0, BigNumber.ROUND_DOWN)
+          break
+        case AssetValueFormat.PRECISION:
+          _value = bnOrZero(params.value)
+            .multipliedBy(bn(10).pow(params.asset ? params.asset.precision : params.precision))
+            .toString()
+          break
+      }
+      this.state = {
+        assetId: params.asset ? params.asset.assetId : params.assetId,
+        precision: params.asset ? params.asset.precision : params.precision,
+        value: _value,
+      }
+    } else {
+      this.loadSerializedAssetValue(params as SerializedAssetValue)
+    }
+  }
+
+  public get assetId(): AssetId {
+    return this.state.assetId
+  }
+
+  public get precision(): number {
+    return this.state.precision
+  }
+
+  public plus(term: AssetValue): AssetValue {
+    this.assertIsSameAsset(term, 'add')
+
+    const value = bnOrZero(this.state.value).plus(bnOrZero(term.toBaseUnit())).toFixed()
+    return new AssetValue({
+      value,
+      assetId: this.state.assetId,
+      precision: this.state.precision,
+      format: AssetValueFormat.BASE_UNIT,
+    })
+  }
+
+  public minus(term: AssetValue): AssetValue {
+    this.assertIsSameAsset(term, 'subtract')
+
+    const value = bnOrZero(this.state.value).minus(bnOrZero(term.toBaseUnit())).toFixed()
+    return new AssetValue({
+      value,
+      assetId: this.state.assetId,
+      precision: this.state.precision,
+      format: AssetValueFormat.BASE_UNIT,
+    })
+  }
+
+  public multipliedBy(factor: string | number): AssetValue {
+    const value = bnOrZero(this.state.value).multipliedBy(bnOrZero(factor)).toFixed()
+    return new AssetValue({
+      value,
+      assetId: this.state.assetId,
+      precision: this.state.precision,
+      format: AssetValueFormat.BASE_UNIT,
+    })
+  }
+
+  public dividedBy(factor: string | number): AssetValue {
+    const value = bnOrZero(this.state.value).dividedBy(bnOrZero(factor)).toFixed()
```

Comment:
I have mentioned staying away from `bnOrZero` specifically for `dividedBy`. Result is `Infinity` vs `NaN` with `toFixed`. Honestly not sure how we handle either value in the client throughout though...